src/sentry/tasks/spans.py

```diff
@@ -1,4 +1,181 @@
+import logging
+import random
+from collections.abc import Mapping, Sequence, Set
+from copy import deepcopy
+from typing import Any
+
+import sentry_sdk
+from sentry_kafka_schemas import get_codec
+from sentry_kafka_schemas.codecs import Codec
+from sentry_kafka_schemas.schema_types.snuba_spans_v1 import SpanEvent
+
+from sentry.event_manager import (
+    Job,
+    ProjectsMapping,
+    _calculate_span_grouping,
+    _detect_performance_problems,
+    _pull_out_data,
+)
+from sentry.issues.grouptype import PerformanceStreamedSpansGroupTypeExperimental
+from sentry.models.project import Project
+from sentry.spans.buffer.redis import RedisSpansBuffer
 from sentry.tasks.base import instrumented_task
+from sentry.utils import metrics
+from sentry.utils.canonical import CanonicalKeyDict
+from sentry.utils.sdk import mark_scope_as_unsafe
+
+SPAN_SCHEMA: Codec[SpanEvent] = get_codec("snuba-spans")
+
+logger = logging.getLogger(__name__)
+
+
+def _deserialize_span(value: bytes) -> Mapping[str, Any]:
+    return SPAN_SCHEMA.decode(value)
+
+
+def build_tree(spans):
+    span_tree = {}
+    root_span_id = None
+
+    for span in spans:
+        span_id = span["span_id"]
+        is_root = span["is_segment"]
+        if is_root:
+            root_span_id = span_id
+        if span_id not in span_tree:
+            span_tree[span_id] = span
+            span_tree[span_id]["children"] = []
+
+    for span in span_tree.values():
+        parent_id = span.get("parent_span_id")
+        if parent_id is not None and parent_id in span_tree:
+            parent_span = span_tree[parent_id]
+            children = parent_span["children"]
+            children.append(span)
+
+    return span_tree, root_span_id
+
+
+def dfs(visited, flattened_spans, tree, span_id):
+    if span_id not in visited:
+        span = deepcopy(tree[span_id])
+        children = span.pop("children")
+
+        visited.add(span_id)
+        flattened_spans.append(span)
+
+        for child in sorted(children, key=lambda span: span["start_timestamp"]):
+            dfs(visited, flattened_spans, tree, child["span_id"])
+
+        tree.pop(span_id)
```

Comment:
I think this could potentially hit max recursion depth if the transaction is just a chain of descendants